deep4cast/models/temporal_cnns.py

```diff
@@ -0,0 +1,108 @@
+# -*- coding: utf-8 -*-
+"""Temporal convolutional neural network module.
+
+This module provides access to temporal CNN models for different time
+series forecasting purposes. These models function with uni- and
+multi-variate time series alike.
+
+"""
+from copy import copy
+from keras.layers import Input, Conv1D, MaxPooling1D, Flatten, Dense, Dropout
+from keras.models import Sequential
+
+
+class TemporalCNN(Sequential):
+    """Extends keras.models.Sequential object.
+
+    Include automatically constructed CNN layers from input topology to
+    make life easier when building regressors and for optimizing network
+    topology.
+
+    :param input_shape: Length and dimesnionality of time series
+    :type input_shape: tuple of ints
+    :param topology: Neural network topology.
+    :type topology: list of integer tuples of length 4
+    :raises: Exception in case of misspecified topolgy.
+
+    """
+
+    def __init__(self, input_shape, topology):
+        """Initialize additional properties."""
+        self._input_shape = input_shape
+        self._topology = topology
+        self._rnd_init = 'glorot_normal'
+        self._activation = 'elu'
+
+        # Check if topology conforms to exepcted model topology kind
+        self.check_topology(topology)
+
+        # Initialize super class with custom layers
+        super(TemporalCNN, self).__init__(self._build_layers())
+
+    def _build_layers(self):
+        """Build CNN model layers.
+
+        Add CNN layers one at a time with layerwise output for easy
+        inspection of internal layers after model is trained.
+
+        """
+        n_layers = len(self._topology)
+        layers = []
+        for i in range(n_layers):
+            n_out, kernel_size, pool_size, pool_stride = self._topology[i]
+            layer_conv = Conv1D(
+                n_out,
+                kernel_size,
+                activation=self._activation,
+                kernel_initializer=self._rnd_init,
+                input_shape=self._input_shape
+            )
+            layer_max = MaxPooling1D(
+                pool_size,
+                pool_stride
+            )
+            layers.append(layer_conv)
+            layers.append(layer_max)
+
+        # Add dense and dropout layers
+        layers.append(Flatten())
+        layers.append(Dense(128, kernel_initializer=self._rnd_init))
```

Comment:
This magic number `128` should be part of topology.

Reply:
True dat :D. I think we should put in some thinking as to how much we would like to handle via a topology object. I did it this way, so that the hyperopt can easily interface with this.

Reply:
Parameter tuning for CNNs is time consuming because of many parameters and nested parameters. 

Ideally, we can just have a model (could be just a handcrafted function in V1) that takes some statistics of the data (input dimension, output dimension, history length, forecast horizon) and outputs the network topology.

Reply:
I've created a new issue #17 to track this.